### ISSUE TITLE
Feature: HTTP RPC API

### ISSUE THREAD

**Issue by cryptoquick** [romanz/electrs#820]
**Is your feature request related to a problem? Please describe.**
We're currently using Blockstream's Esplora with BDK, but it's quite heavy and out of date with current electrs.

**Describe the solution you'd like**
I've considered making a crate called `electrs-http-proxy` that's just a reimplementation of the Esplora REST API that proxies to wrap electrs. Would this make sense to do as a separate crate?

**Describe alternatives you've considered**
An HTTP server will likely be opinionated and might complicate maintenance of electrs itself, and I'd be open to other implementations behind feature gating, but if I was doing it, I'd probably default to updating the interface to use axum, while referencing the hyper handler interface used by Blockstream.

**Additional context**
For context, a link to the Blockstream fork for Esplora:
https://github.com/Blockstream/electrs

Their REST request handler, using hyper:
https://github.com/Blockstream/electrs/blob/a808b51d0d9301fa82390b985c57551966001f9b/src/rest.rs#L615

I'm open to opinions, advice, especially of anything I might seem to be unaware of.

**Comment by romanz** [comment]
Good idea, a new proxy crate would be a great contribution!

**Comment by stevenroose** [comment]
Did this happen?

**Comment by romanz** [comment]
Unfortunately, no :(

**Comment by Kixunil** [comment]
@stevenroose I've posted a jsonrpc<->websocket proxy somewhere, you can try to find it.

**Comment by darioAnongba** [comment]
Hi all,

This feature missing seems problematic to me because of the following:
- Most users running their own infra on Start9 or Umbrel install `electrs`.
- Running Esplora on a personal server is too heavy even with the `light-mode`.
- BDK doesn't support have support for a websocket client (could be added but there is no specification for it), nor does this websocket proxy exist as an app on Umbrel or Start9.

This renders the connection to your own node very difficult from browser or browser extensions. The reason this is problematic is because browsers are becoming more and more akin to full-fledge OSs like we see on the Chromebooks and browser extensions might become popular for Bitcoin in Lightning wallets.

 In conclusion, I think we need a common effort between Blockstream, mempool and electrum clients to converge on a standard API for indexers. There are even other projects like Blockbook or Bitcore that have different APIs (also not aimed at personal servers). 
Your opinion?

**Comment by Kixunil** [comment]
> * Most users running their own infra on Start9 or Umbrel install `electrs`.

Right there. If they can package several services and make sure they communicate together correctly, they can also package one more proxy service. I see no reason to pile that into an **Electrum protocol** implementation, especially because the thing already exists and is reasonable. It's not rocket science. I'm doing that kind of packaging at [CADR](https://github.com/debian-cryptoanarchy/cryptoanarchy-deb-repo-builder/) and I don't see why it should take more than a few tens of minutes to get a first draft ready. This is also a proof that I'm not just throwing integrators under the bus, since I'm one of them. :)

> * there is no specification for it

Neither is there a specification of Electrum over HTTP, Electrum is only specified over TCP. So if we have to specify something anyway, do it with websocket, which is clearly superior because it supports notifications which are critical for performance and it's trivial to map it 1:1 to the existing Json RPC.
Also note that in either case it's critical to specify some authentication scheme because `electrs` has no DoS protection (for a good reason).

> * nor does this websocket proxy exist as an app on Umbrel or Start9.

You can go and make a PR to them. I heard they are Open Source. I'll be happy to accept a PR to CADR as well, though I will likely be able to do it myself once it's specified.

> This renders the connection to your own node very difficult from browser or browser extensions.

If you want to see what's "very difficult" come develop rust-bitcoin. This is cakewalk.

> I think we need a common effort between Blockstream, mempool and electrum servers to converge on a standard API for indexers.

We don't. The proxy trivially works for all of them out-of-the-box as long as they continue to support the already-standard Electrum protocol. (IIRC Blockstream fork doesn't).
